cel/inlining_test.go

```diff
@@ -0,0 +1,577 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cel_test
+
+import (
+	"testing"
+
+	"github.com/google/cel-go/cel"
+
+	proto3pb "github.com/google/cel-go/test/proto3pb"
+)
+
+func TestInliningOptimizer(t *testing.T) {
+	type varExpr struct {
+		name  string
+		alias string
+		t     *cel.Type
+		expr  string
+	}
+	tests := []struct {
+		expr    string
+		vars    []varExpr
+		inlined string
+		folded  string
+	}{
+		{
+			expr: `a || b`,
+			vars: []varExpr{
+				{
+					name: "a",
+					t:    cel.BoolType,
+				},
+				{
+					name:  "b",
+					alias: "bravo",
+					t:     cel.BoolType,
+					expr:  `'hello'.contains('lo')`,
+				},
+			},
+			inlined: `a || "hello".contains("lo")`,
+			folded:  `true`,
+		},
+		{
+			expr: `a + [a]`,
+			vars: []varExpr{
+				{
+					name:  "a",
+					alias: "alpha",
+					t:     cel.DynType,
+					expr:  `dyn([1, 2])`,
+				},
+			},
+			inlined: `cel.bind(alpha, dyn([1, 2]), alpha + [alpha])`,
+			folded:  `[1, 2, [1, 2]]`,
+		},
+		{
+			expr: `a && (a || b)`,
+			vars: []varExpr{
+				{
+					name:  "a",
+					alias: "alpha",
+					t:     cel.BoolType,
+					expr:  `'hello'.contains('lo')`,
+				},
+				{
+					name: "b",
+					t:    cel.BoolType,
+				},
+			},
+			inlined: `cel.bind(alpha, "hello".contains("lo"), alpha && (alpha || b))`,
+			folded:  `true`,
+		},
+		{
+			expr: `a && b && a`,
+			vars: []varExpr{
+				{
+					name:  "a",
+					alias: "alpha",
+					t:     cel.BoolType,
+					expr:  `'hello'.contains('lo')`,
+				},
+				{
+					name: "b",
+					t:    cel.BoolType,
+				},
+			},
+			inlined: `cel.bind(alpha, "hello".contains("lo"), alpha && b && alpha)`,
+			folded:  `cel.bind(alpha, true, alpha && b && alpha)`,
```

Comment:
Shouldn't this one fold to just `b`? I guess the constant folding might have to be augmented to better handle binds (ironically, I think just choosing to not bind for this example would rightly produce `b`).

Reply:
Just noticed that we might actually not be folding simple literals correctly -- sent https://github.com/google/cel-go/pull/833

Reply:
This one should fold to `b`, but the constant folding isn't quite smart enough to reason about the `bind` logic to determine that the expression could be represented as `alpha && b && alpha`. Specifically, because `b` appears and it's not known, the bind() doesn't fold.